# Car Listing Project

The aim of this project is to clean the data and analyze the included used car listings

In [61]:
import pandas as pd
import numpy as np
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')


In [62]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [63]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [64]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

so far we can see that a few columns have null values and that there are string and integer columns, mostly strings though


In [65]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [66]:
new_cols = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gear_box', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']


In [67]:
autos.columns = new_cols

In [68]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

so far we have converted the column names from camelcase to snakecase

In [69]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-11 22:38:16  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year gear_box      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

nr_of pictures seems like a column with only 0 values. 

power_PS seems like a column with really large standard deviation and outliers.


In [70]:
autos.loc[:,'price'].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

we can see from the unique values, we need to replace '$' and ','

In [71]:
autos.loc[:,'price'] = autos.loc[:,'price'].str.replace('$','')
autos.loc[:,'price'] = autos.loc[:,'price'].str.replace(',','')
autos.loc[:,'price'] = autos.loc[:,'price'].astype(int)

In [72]:
autos.loc[:,'odometer'].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [73]:
autos.loc[:,'odometer']=autos.loc[:,'odometer'].str.replace('km','')
autos.loc[:,'odometer']=autos.loc[:,'odometer'].str.replace(',','')

In [74]:
autos.loc[:,'odometer'] = autos.loc[:,'odometer'].astype(int)

In [75]:
autos.rename({'odometer':'odometer_km'},axis=1,inplace=True)

In [76]:
autos.loc[:,['price','odometer_km']].describe()

price    odometer_km
count  5.000000e+04   50000.000000
mean   9.840044e+03  125732.700000
std    4.811044e+05   40042.211706
min    0.000000e+00    5000.000000
25%    1.100000e+03  125000.000000
50%    2.950000e+03  150000.000000
75%    7.200000e+03  150000.000000
max    1.000000e+08  150000.000000

In [77]:
autos.loc[:,'price'].unique().shape[0] #see how many unique values

2357

In [78]:
autos.loc[:,'price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [79]:
autos.loc[:,'price'].value_counts().sort_index(ascending=False).head(50)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
145000      1
139997      1
137999      1
135000      1
130000      1
129000      1
128000      1
120000      2
119900      1
119500      1
116000      1
115991      1
115000      1
114400      1
109999      1
105000      2
104900      1
99900       2
99000       2
98500       1
Name: price, dtype: int64

We have so far sorted by index number (values)

There are a number of listings with prices below \$30, including about 1,500 at \$0. There are also a small number of listings with very high values, including 14 at around or over $1 million.

Given that eBay is an auction site, there could legitimately be items where the opening bid is \$1. We will keep the \$1 items, but remove anything above \$350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

In [80]:
#only keep non_outlier prices rows

In [81]:
autos = autos[autos.loc[:,'price'] < 350000]

In [82]:
autos.describe()

price  registration_year      power_PS    odometer_km  \
count   49985.000000       49985.000000  49985.000000   49985.000000   
mean     5714.637531        2005.075503    116.333520  125738.921677   
std      8850.742575         105.728207    209.213067   40034.891684   
min         0.000000        1000.000000      0.000000    5000.000000   
25%      1100.000000        1999.000000     70.000000  125000.000000   
50%      2950.000000        2003.000000    105.000000  150000.000000   
75%      7200.000000        2008.000000    150.000000  150000.000000   
max    345000.000000        9999.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        49985.000000         49985.0  49985.000000  
mean             5.723777             0.0  50812.410383  
std              3.711856             0.0  25777.512429  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30451.000000  
50%              6.000000             0.0  49565.000000  
75%              9.000000             0.0  71522.000000  
max             12.000000             0.0  99998.000000

In [83]:
autos.loc[:,'date_crawled'].value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2016-03-05 14:06:30    0.00002
2016-03-05 14:06:40    0.00002
2016-03-05 14:07:04    0.00002
2016-03-05 14:07:08    0.00002
2016-03-05 14:07:21    0.00002
2016-03-05 14:07:26    0.00002
2016-03-05 14:07:40    0.00002
2016-03-05 14:07:45    0.00002
2016-03-05 14:08:00    0.00004
2016-03-05 14:08:05    0.00004
2016-03-05 14:08:27    0.00002
2016-03-05 14:08:42    0.00002
2016-03-05 14:09:02    0.00004
2016-03-05 14:09:05    0.00002
2016-03-05 14:09:20    0.00002
2016-03-05 14:09:22    0.00002
2016-03-05 14:09:38    0.00002
2016-03-05 14:09:46    0.00002
2016-03-05 14:09:56    0.00002
2016-03-05 14:09:57    0.00002
2016-03-05 14:09:58    0.00004
2016-03-05 14:10:18    0.00002
2016-03-05 14:10:20    0.00002
2016-03-05 14:10:46    0.00002
2016-03-05 14:11:03    0.00002
2016-03-05 14:11:05    0.00002
2016-03-05 14:11:14    0.00002
2016-03-05 14:11:15    0.00002
2016-03-05 14:11:25    0.00002
2016-03-05 14:11:40    0.00002
                        ...   
2016-04-07 10:36:19    0.00002
2016-04-

we just got the relative frequency in ascending order of time 

we observe here that time values seem to be pretty evenly distributed here. Looks like the site was crawled daily over roughly a one month period in March and April 2016

In [84]:
autos.loc[:,'ad_created'].value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2015-06-11 00:00:00    0.000020
2015-08-10 00:00:00    0.000020
2015-09-09 00:00:00    0.000020
2015-11-10 00:00:00    0.000020
2015-12-05 00:00:00    0.000020
2015-12-30 00:00:00    0.000020
2016-01-03 00:00:00    0.000020
2016-01-07 00:00:00    0.000020
2016-01-10 00:00:00    0.000040
2016-01-13 00:00:00    0.000020
2016-01-14 00:00:00    0.000020
2016-01-16 00:00:00    0.000020
2016-01-22 00:00:00    0.000020
2016-01-27 00:00:00    0.000060
2016-01-29 00:00:00    0.000020
2016-02-01 00:00:00    0.000020
2016-02-02 00:00:00    0.000040
2016-02-05 00:00:00    0.000040
2016-02-07 00:00:00    0.000020
2016-02-08 00:00:00    0.000020
2016-02-09 00:00:00    0.000040
2016-02-11 00:00:00    0.000020
2016-02-12 00:00:00    0.000060
2016-02-14 00:00:00    0.000040
2016-02-16 00:00:00    0.000020
2016-02-17 00:00:00    0.000020
2016-02-18 00:00:00    0.000040
2016-02-19 00:00:00    0.000060
2016-02-20 00:00:00    0.000040
2016-02-21 00:00:00    0.000060
                         ...   
2016-03-

for ad_created we observe that the earlier times occur a lot more frequently and later times occur far fewer times. There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

this part is giving us a good idea of the generall distribution of each of the time columns

In [85]:
autos.loc[:,'last_seen'].value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2016-03-05 14:45:46    0.00002
2016-03-05 14:46:02    0.00002
2016-03-05 14:49:34    0.00002
2016-03-05 15:16:11    0.00002
2016-03-05 15:16:47    0.00002
2016-03-05 15:28:10    0.00002
2016-03-05 15:41:30    0.00002
2016-03-05 15:45:43    0.00002
2016-03-05 15:47:38    0.00002
2016-03-05 15:47:44    0.00002
2016-03-05 16:45:57    0.00002
2016-03-05 16:47:28    0.00002
2016-03-05 17:15:45    0.00002
2016-03-05 17:16:12    0.00002
2016-03-05 17:16:14    0.00002
2016-03-05 17:16:23    0.00002
2016-03-05 17:17:02    0.00002
2016-03-05 17:39:19    0.00002
2016-03-05 17:40:14    0.00002
2016-03-05 17:44:50    0.00002
2016-03-05 17:44:54    0.00002
2016-03-05 17:46:01    0.00002
2016-03-05 18:17:58    0.00002
2016-03-05 18:47:14    0.00002
2016-03-05 18:50:38    0.00002
2016-03-05 19:15:08    0.00002
2016-03-05 19:15:20    0.00002
2016-03-05 19:15:42    0.00002
2016-03-05 19:16:36    0.00002
2016-03-05 19:17:17    0.00002
                        ...   
2016-04-07 14:58:09    0.00004
2016-04-

In [86]:
last_seen is also pretty evenly distributed


SyntaxError: invalid syntax (<ipython-input-86-529c32fbeb41>, line 1)

In [87]:
autos["registration_year"].describe()


count    49985.000000
mean      2005.075503
std        105.728207
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.



Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate

In [88]:
autos.loc[:,'registration_year'].value_counts().sort_index().head(30)

1000     1
1001     1
1111     1
1500     1
1800     2
1910     9
1927     1
1929     1
1931     1
1934     2
1937     4
1938     1
1939     1
1941     2
1943     1
1948     1
1950     3
1951     2
1952     1
1953     1
1954     2
1955     2
1956     5
1957     2
1958     4
1959     7
1960    33
1961     6
1962     4
1963     9
Name: registration_year, dtype: int64

It seems that in 1910, 9 cars were registered, that seems like a good earliest value allowed for this columns

In [89]:
autos.loc[:,'registration_year'].value_counts().sort_index(ascending=False).head(15)

9999       4
9996       1
9000       2
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       3
2018     491
2017    1452
2016    1315
Name: registration_year, dtype: int64

It seems that 2016 is a good max value allowed here

In [90]:
autos = autos[autos.loc[:,'registration_year'].between(1900,2016)]

In [91]:
autos.describe()

price  registration_year      power_PS    odometer_km  \
count   48015.000000       48015.000000  48015.000000   48015.000000   
mean     5804.348599        2002.805727    117.046402  125551.598459   
std      8966.172382           7.306040    195.138824   40098.668741   
min         0.000000        1910.000000      0.000000    5000.000000   
25%      1150.000000        1999.000000     71.000000  100000.000000   
50%      2990.000000        2003.000000    107.000000  150000.000000   
75%      7399.000000        2008.000000    150.000000  150000.000000   
max    345000.000000        2016.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        48015.000000         48015.0  48015.000000  
mean             5.768093             0.0  50934.821535  
std              3.696675             0.0  25790.585059  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30459.000000  
50%              6.000000             0.0  49692.000000  
75%              9.000000             0.0  71665.000000  
max             12.000000             0.0  99998.000000

In [92]:
common_brands = autos.loc[:,'brand'].value_counts().head(20).index
#getting a list of the 20 most frequent car brands

In [93]:
mean_price_dict = {}
for brand in common_brands:
    brand_data = autos[autos.loc[:,'brand']==brand]
    mean = brand_data.loc[:,'price'].mean()
    mean_price_dict[brand] = int(mean)
    
mean_price_dict
#getting the mean price for the 20 most common car brands

{'audi': 9093,
 'bmw': 8102,
 'citroen': 3699,
 'fiat': 2711,
 'ford': 3652,
 'hyundai': 5308,
 'mazda': 4010,
 'mercedes_benz': 8485,
 'mini': 10460,
 'nissan': 4664,
 'opel': 2876,
 'peugeot': 3039,
 'renault': 2395,
 'seat': 4296,
 'skoda': 6334,
 'smart': 3542,
 'sonstige_autos': 10805,
 'toyota': 5115,
 'volkswagen': 5231,
 'volvo': 4757}

we aggregated across brands to understand mean price. We observed that in the top 6 brands, there's a distinct price gap.

Audi, BMW and Mercedes Benz are more expensive
Ford and Opel are less expensive
Volkswagen is in between

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. 

In [94]:
mean_mileage_dict ={}
for brand in common_brands:
    brand_data = autos[autos.loc[:,'brand']==brand]
    mean = brand_data.loc[:,'odometer_km'].mean()
    mean_mileage_dict[brand] = mean
    

mean_mileage_dict

{'audi': 129287.78018799711,
 'bmw': 132431.38368351315,
 'citroen': 119580.8383233533,
 'fiat': 116553.94524959743,
 'ford': 124068.65671641791,
 'hyundai': 106511.62790697675,
 'mazda': 124745.5295735901,
 'mercedes_benz': 130856.0821139987,
 'mini': 88602.40963855422,
 'nissan': 118572.41379310345,
 'opel': 129223.14208702349,
 'peugeot': 127136.81241184767,
 'renault': 128183.81706244503,
 'seat': 121563.57388316152,
 'skoda': 110954.54545454546,
 'smart': 99595.80838323354,
 'sonstige_autos': 87466.53919694072,
 'toyota': 115709.51585976628,
 'volkswagen': 128724.10407461954,
 'volvo': 138355.85585585586}

In [96]:
mean_mileage = pd.Series(mean_mileage_dict).sort_values(ascending=False)
mean_prices = pd.Series(mean_price_dict).sort_values(ascending=False)

In [100]:
brand_mileage = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_mileage

mean_mileage
volvo           138355.855856
bmw             132431.383684
mercedes_benz   130856.082114
audi            129287.780188
opel            129223.142087
volkswagen      128724.104075
renault         128183.817062
peugeot         127136.812412
mazda           124745.529574
ford            124068.656716
seat            121563.573883
citroen         119580.838323
nissan          118572.413793
fiat            116553.945250
toyota          115709.515860
skoda           110954.545455
hyundai         106511.627907
smart            99595.808383
mini             88602.409639
sonstige_autos   87466.539197

In [101]:
brand_price = pd.DataFrame(mean_prices,columns=['mean_price'])
brand_price

mean_price
sonstige_autos       10805
mini                 10460
audi                  9093
mercedes_benz         8485
bmw                   8102
skoda                 6334
hyundai               5308
volkswagen            5231
toyota                5115
volvo                 4757
nissan                4664
seat                  4296
mazda                 4010
citroen               3699
ford                  3652
smart                 3542
peugeot               3039
opel                  2876
fiat                  2711
renault               2395

In [102]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
volvo           138355.855856        4757
bmw             132431.383684        8102
mercedes_benz   130856.082114        8485
audi            129287.780188        9093
opel            129223.142087        2876
volkswagen      128724.104075        5231
renault         128183.817062        2395
peugeot         127136.812412        3039
mazda           124745.529574        4010
ford            124068.656716        3652
seat            121563.573883        4296
citroen         119580.838323        3699
nissan          118572.413793        4664
fiat            116553.945250        2711
toyota          115709.515860        5115
skoda           110954.545455        6334
hyundai         106511.627907        5308
smart            99595.808383        3542
mini             88602.409639       10460
sonstige_autos   87466.539197       10805


The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage